File: src/MainWindow.cpp
#include "MainWindow.h"

#include "MetadataService.h"
#include "TagEditorWidget.h"
#include "TrackTableModel.h"

#include <QAction>
#include <QApplication>
#include <QDesktopServices>
#include <QDate>
#include <QDragEnterEvent>
#include <QDragMoveEvent>
#include <QDropEvent>
#include <QFileDialog>
#include <QFileInfo>
#include <QDir>
#include <QGuiApplication>
#include <QHeaderView>
#include <QAbstractItemView>
#include <QMessageBox>
#include <QPainter>
#include <QEvent>
#include <QMimeData>
#include <QSplitter>
#include <QItemSelectionModel>
#include <QStatusBar>
#include <QTableView>
#include <QToolBar>
#include <QMenuBar>
#include <QMenu>
#include <QSettings>
#include <QDirIterator>
#include <QUrl>
#include <QVBoxLayout>
#include <QWidget>
#include <QSize>

#ifdef Q_OS_WIN
#include <windows.h>
#include <shellapi.h>
#endif

#ifdef Q_OS_MAC
#include <QProcess>
#endif

MainWindow::MainWindow(QWidget *parent)
    : QMainWindow(parent)
{
    m_metadataService = new MetadataService;
    setupUi();
    setupActions();
    setupToolbar();
    setupMenuBar();
    updateSelectionState();
}

void MainWindow::setupUi()
{
    auto *central = new QWidget;
    auto *layout = new QVBoxLayout(central);

    auto *splitter = new QSplitter(Qt::Vertical, central);

    m_tableView = new QTableView(splitter);
    m_tableView->setAcceptDrops(true);
    m_tableView->viewport()->setAcceptDrops(true);
    m_tableView->setDragEnabled(false);
    m_tableView->setDragDropMode(QAbstractItemView::DropOnly);
    m_tableView->setDropIndicatorShown(true);
    m_tableView->setDefaultDropAction(Qt::CopyAction);
    m_tableView->setSelectionBehavior(QAbstractItemView::SelectRows);
    m_tableView->setSelectionMode(QAbstractItemView::SingleSelection);
    m_tableView->setAlternatingRowColors(true);
    m_tableView->setSortingEnabled(false);
    m_tableView->viewport()->installEventFilter(this);
    m_tableView->setContextMenuPolicy(Qt::CustomContextMenu);
    connect(m_tableView, &QWidget::customContextMenuRequested, this, &MainWindow::onShowContextMenu);

    // Create and install the model before configuring header sections. If the
    // header is configured before a model is set, calls that resolve logical
    // <-> visual indices can return -1 and trigger assertions in Qt.
    m_model = new TrackTableModel(this);
    m_tableView->setModel(m_model);

    // Ensure icons (cover art) are visible at a reasonable size
    m_tableView->setIconSize(QSize(48, 48));
    m_tableView->verticalHeader()->setDefaultSectionSize(56);

    m_tableView->verticalHeader()->setVisible(false);
    m_tableView->horizontalHeader()->setStretchLastSection(true);
    m_tableView->horizontalHeader()->setSectionResizeMode(0, QHeaderView::Fixed);
    m_tableView->horizontalHeader()->resizeSection(0, 64);
    m_tableView->horizontalHeader()->setSectionResizeMode(1, QHeaderView::Interactive);
    m_tableView->horizontalHeader()->setSectionResizeMode(2, QHeaderView::Interactive);
    m_tableView->horizontalHeader()->setSectionResizeMode(3, QHeaderView::Interactive);
    m_tableView->horizontalHeader()->setSectionResizeMode(4, QHeaderView::Interactive);
    m_tableView->horizontalHeader()->setSectionResizeMode(5, QHeaderView::Interactive);
    m_tableView->horizontalHeader()->setMinimumSectionSize(48);

    m_editor = new TagEditorWidget(splitter);
    m_editor->setEditingEnabled(false);

    splitter->addWidget(m_tableView);
    splitter->addWidget(m_editor);
    splitter->setStretchFactor(0, 3);
    splitter->setStretchFactor(1, 2);
    splitter->setChildrenCollapsible(false);

    layout->addWidget(splitter);
    setCentralWidget(central);

    connect(m_tableView->selectionModel(), &QItemSelectionModel::currentRowChanged,
            this, &MainWindow::onSelectionChanged);
    connect(m_editor, &TagEditorWidget::saveRequested,
            this, &MainWindow::onSaveRequested);

    statusBar()->showMessage(tr("Load audio files to begin."));
    resize(1300, 900);
    setWindowTitle(tr("TagTune"));
}

void MainWindow::setupActions()
{
    m_openAction = new QAction(tr("Open Files..."), this);
    m_openFolderAction = new QAction(tr("Open Folder..."), this);
    m_saveAction = new QAction(tr("Save Current"), this);
    m_saveAction->setShortcut(QKeySequence::Save);
    m_clearAction = new QAction(tr("Clear Selection"), this);
    m_removeAction = new QAction(tr("Remove"), this);
    m_showInFinderAction = new QAction(tr("Show in Finder/Explorer"), this);
    m_recentMenu = nullptr;

    connect(m_openAction, &QAction::triggered, this, &MainWindow::loadFiles);
    connect(m_openFolderAction, &QAction::triggered, this, &MainWindow::loadFolderDialog);
    connect(m_saveAction, &QAction::triggered, this, [this]() {
        if (!m_tableView->currentIndex().isValid()) {
            return;
        }
        onSaveRequested(m_editor->track());
    });
    connect(m_clearAction, &QAction::triggered, this, [this]() {
        m_tableView->clearSelection();
        m_editor->clearTrack();
        updateSelectionState();
        statusBar()->showMessage(tr("Selection cleared."), 2000);
    });

    connect(m_removeAction, &QAction::triggered, this, &MainWindow::onRemoveRequested);
    connect(m_showInFinderAction, &QAction::triggered, this, &MainWindow::onShowInFinderRequested);
}

void MainWindow::setupToolbar()
{
#ifdef Q_OS_MAC
    // On macOS prefer the native menu bar for app-level actions (avoid duplicate top bars)
    // Keep a minimal in-window UI; do not add the top toolbar here so the global menu is primary.
#else
    auto *toolbar = addToolBar(tr("Main"));
    toolbar->setMovable(false);
    toolbar->addAction(m_openAction);
    toolbar->addAction(m_saveAction);
    toolbar->addAction(m_clearAction);
    toolbar->addAction(m_removeAction);
    toolbar->addSeparator();
#endif
}

void MainWindow::setupMenuBar()
{
    // Menu bar with File menu and Recent submenu (on macOS the QMainWindow menu bar will be native)
    QMenuBar *mb = menuBar();
    // On macOS use the native (global) menu bar
    mb->setNativeMenuBar(true);
    QMenu *fileMenu = mb->addMenu(tr("File"));

    fileMenu->addAction(m_openAction);
    fileMenu->addAction(m_openFolderAction);

    fileMenu->addSeparator();

    fileMenu->addAction(m_saveAction);

    fileMenu->addAction(m_clearAction);
    fileMenu->addAction(m_removeAction);
    fileMenu->addAction(m_showInFinderAction);

    fileMenu->addSeparator();

    // Recent menu (populated from QSettings)
    m_recentMenu = new QMenu(tr("Open Recent"), this);
    fileMenu->addMenu(m_recentMenu);
    updateRecentMenu();

    // Quit on Mac uses standard role
    QAction *quitAct = fileMenu->addAction(tr("Quit"));
    quitAct->setMenuRole(QAction::QuitRole);
    connect(quitAct, &QAction::triggered, qApp, &QCoreApplication::quit);
}

void MainWindow::updateRecentMenu()
{
    m_recentMenu->clear();
    QSettings s;
    const QStringList recent = s.value("recentPaths").toStringList();
    if (recent.isEmpty()) {
        m_recentMenu->setEnabled(false);
        return;
    }
    m_recentMenu->setEnabled(true);
    for (const QString &p : recent) {
        QAction *act = m_recentMenu->addAction(p);
        connect(act, &QAction::triggered, this, [this, p]() {
            QFileInfo fi(p);
            if (fi.exists() && fi.isDir()) {
                loadFolder(p);
            } else if (fi.exists() && fi.isFile()) {
                // Avoid duplicates when opening from recent
                QString key = fi.canonicalFilePath();
                if (key.isEmpty()) key = fi.absoluteFilePath();
                if (m_loadedPaths.contains(key)) return;
                const QVector<Track> loaded = m_metadataService->loadTracks(QStringList{p});
                for (const Track &t : loaded) {
                    QFileInfo tfi(t.filePath);
                    QString tkey = tfi.canonicalFilePath();
                    if (tkey.isEmpty()) tkey = tfi.absoluteFilePath();
                    if (m_loadedPaths.contains(tkey)) continue;
                    m_model->addTrack(t);
                    m_loadedPaths.insert(tkey);
                }
                if (!m_tableView->currentIndex().isValid()) m_tableView->selectRow(0);
            } else {
                QMessageBox::warning(this, tr("TagTune"), tr("That recent location no longer exists:\n%1").arg(p));
                removeFromRecent(p);
            }
        });
    }
}

void MainWindow::addToRecent(const QString &path)
{
    QSettings s;
    QStringList recent = s.value("recentPaths").toStringList();
    recent.removeAll(path);
    recent.prepend(path);
    while (recent.size() > 12) recent.removeLast();
    s.setValue("recentPaths", recent);
    updateRecentMenu();
}

void MainWindow::removeFromRecent(const QString &path)
{
    QSettings s;
    QStringList recent = s.value("recentPaths").toStringList();
    recent.removeAll(path);
    s.setValue("recentPaths", recent);
    updateRecentMenu();
}

static bool isAudioFileName(const QString &name)
{
    const QString n = name.toLower();
    return n.endsWith(".mp3") || n.endsWith(".m4a") || n.endsWith(".mp4") || n.endsWith(".flac") || n.endsWith(".wav") || n.endsWith(".ogg") || n.endsWith(".aac");
}

void MainWindow::loadFolderDialog()
{
    const QString folder = QFileDialog::getExistingDirectory(this, tr("Select Music Folder"));
    if (folder.isEmpty()) return;
    addToRecent(folder);
    loadFolder(folder);
}

void MainWindow::loadFolder(const QString &folderPath)
{
    QVector<QString> files;
    QDirIterator it(folderPath, QDir::Files, QDirIterator::Subdirectories);
    while (it.hasNext()) {
        const QString f = it.next();
        if (isAudioFileName(f)) files.append(f);
    }

    if (files.isEmpty()) {
        QMessageBox::information(this, tr("TagTune"), tr("No audio files found in %1").arg(folderPath));
        return;
    }

    QStringList qsl;
    for (const QString &s : files) qsl.append(s);
    const QVector<Track> loadedTracks = m_metadataService->loadTracks(qsl);
    int added = 0;
    for (const Track &track : loadedTracks) {
        QFileInfo tfi(track.filePath);
        QString key = tfi.canonicalFilePath();
        if (key.isEmpty()) key = tfi.absoluteFilePath();
        if (m_loadedPaths.contains(key)) continue;
        m_model->addTrack(track);
        m_loadedPaths.insert(key);
        ++added;
    }
    if (!m_tableView->currentIndex().isValid() && added > 0) m_tableView->selectRow(0);
    refreshEditorFromSelection();
    statusBar()->showMessage(tr("Loaded %1 file(s) from folder.").arg(added), 3000);
}

// Demo data removed: application no longer seeds sample tracks.

void MainWindow::loadFiles()
{
    const QStringList files = QFileDialog::getOpenFileNames(
        this,
        tr("Open Audio Files"),
        QString(),
        tr("Audio Files (*.mp3 *.flac *.wav *.ogg *.m4a *.aac);;All Files (*.*)")
    );

    if (files.isEmpty()) {
        return;
    }
    addToRecent(QFileInfo(files.first()).absolutePath());
    loadPaths(files);
}

void MainWindow::loadPaths(const QStringList &paths)
{
    QStringList toLoad;
    toLoad.reserve(paths.size());
    for (const QString &path : paths) {
        QFileInfo fi(path);
        if (!fi.exists() || !fi.isFile() || !isAudioFileName(fi.fileName())) {
            continue;
        }
        QString key = fi.canonicalFilePath();
        if (key.isEmpty()) key = fi.absoluteFilePath();
        if (!m_loadedPaths.contains(key)) {
            toLoad.append(path);
        }
    }

    if (toLoad.isEmpty()) {
        statusBar()->showMessage(tr("No new audio files to load."), 2000);
        return;
    }

    const QVector<Track> loadedTracks = m_metadataService->loadTracks(toLoad);
    int added = 0;
    for (const Track &track : loadedTracks) {
        QFileInfo tfi(track.filePath);
        QString key = tfi.canonicalFilePath();
        if (key.isEmpty()) key = tfi.absoluteFilePath();
        if (m_loadedPaths.contains(key)) continue;
        m_model->addTrack(track);
        m_loadedPaths.insert(key);
        ++added;
    }

    if (!m_tableView->currentIndex().isValid() && added > 0) {
        m_tableView->selectRow(0);
    }

    refreshEditorFromSelection();
    statusBar()->showMessage(tr("Loaded %1 file(s)." ).arg(added), 3000);
}

void MainWindow::onSelectionChanged()
{
    refreshEditorFromSelection();
}

void MainWindow::updateSelectionState()
{
    const bool hasSelection = m_tableView && m_tableView->currentIndex().isValid();
    m_editor->setEditingEnabled(hasSelection);
    m_saveAction->setEnabled(hasSelection);
    m_clearAction->setEnabled(hasSelection);
    m_removeAction->setEnabled(hasSelection);
    m_showInFinderAction->setEnabled(hasSelection);
}

void MainWindow::refreshEditorFromSelection()
{
    const QModelIndex current = m_tableView->currentIndex();
    if (!current.isValid()) {
        m_editor->clearTrack();
        updateSelectionState();
        return;
    }

    m_editor->setTrack(m_model->trackAt(current.row()));
    updateSelectionState();
}

void MainWindow::onSaveRequested(const Track &track)
{
    const QModelIndex current = m_tableView->currentIndex();
    if (!current.isValid()) {
        return;
    }

    QString errorMessage;
    if (!m_metadataService->saveTrack(track, &errorMessage)) {
        QMessageBox::warning(this, tr("TagTune"), errorMessage);
        return;
    }

    m_model->updateTrack(current.row(), track);
    statusBar()->showMessage(tr("Saved tags for %1").arg(track.title.isEmpty() ? QFileInfo(track.filePath).baseName() : track.title), 3000);
}

void MainWindow::onRemoveRequested()
{
    const QModelIndex current = m_tableView->currentIndex();
    if (!current.isValid()) return;

    const int row = current.row();
    const Track t = m_model->trackAt(row);

    // Remove from loaded paths
    QFileInfo fi(t.filePath);
    QString key = fi.canonicalFilePath();
    if (key.isEmpty()) key = fi.absoluteFilePath();
    m_loadedPaths.remove(key);

    m_model->removeTrack(row);

    // update selection
    const int rows = m_model->rowCount();
    if (rows > 0) {
        m_tableView->selectRow(qMin(row, rows - 1));
    } else {
        m_editor->clearTrack();
    }
    refreshEditorFromSelection();
    updateSelectionState();

    statusBar()->showMessage(tr("Removed %1").arg(fi.fileName()), 2000);
}

QString MainWindow::selectedTrackPath() const
{
    const QModelIndex current = m_tableView->currentIndex();
    if (!current.isValid()) return {};
    return m_model->trackAt(current.row()).filePath;
}

void MainWindow::revealInFileManager(const QString &path) const
{
    QFileInfo fi(path);
    if (!fi.exists()) {
        QMessageBox::warning(const_cast<MainWindow *>(this), tr("TagTune"), tr("File no longer exists:\n%1").arg(path));
        return;
    }

    const QString absolutePath = fi.absoluteFilePath();
#ifdef Q_OS_MAC
    QProcess::startDetached(QStringLiteral("open"), {QStringLiteral("-R"), absolutePath});
#elif defined(Q_OS_WIN)
    const QString native = QDir::toNativeSeparators(absolutePath);
    const QString args = QStringLiteral("/select,\"") + native + QStringLiteral("\"");
    const std::wstring argsW = args.toStdWString();
    ShellExecuteW(nullptr, L"open", L"explorer.exe", argsW.c_str(), nullptr, SW_SHOW);
#else
    QDesktopServices::openUrl(QUrl::fromLocalFile(fi.absolutePath()));
#endif
}

void MainWindow::onShowInFinderRequested()
{
    const QString path = selectedTrackPath();
    if (path.isEmpty()) return;
    revealInFileManager(path);
}

void MainWindow::onShowContextMenu(const QPoint &pos)
{
    const QModelIndex idx = m_tableView->indexAt(pos);
    QMenu menu(this);
    if (idx.isValid()) {
        menu.addAction(m_removeAction);
        menu.addAction(m_showInFinderAction);
    }
    menu.exec(m_tableView->viewport()->mapToGlobal(pos));
}

bool MainWindow::eventFilter(QObject *watched, QEvent *event)
{
    if ((watched == m_tableView || watched == m_tableView->viewport()) && event) {
        if (event->type() == QEvent::DragEnter) {
            auto *dragEvent = static_cast<QDragEnterEvent *>(event);
            if (dragEvent->mimeData()->hasUrls()) {
                dragEvent->acceptProposedAction();
                return true;
            }
        } else if (event->type() == QEvent::DragMove) {
            auto *dragEvent = static_cast<QDragMoveEvent *>(event);
            if (dragEvent->mimeData()->hasUrls()) {
                dragEvent->acceptProposedAction();
                return true;
            }
        } else if (event->type() == QEvent::Drop) {
            auto *dropEvent = static_cast<QDropEvent *>(event);
            QStringList paths;
            for (const QUrl &url : dropEvent->mimeData()->urls()) {
                if (url.isLocalFile()) paths.append(url.toLocalFile());
            }
            loadPaths(paths);
            dropEvent->acceptProposedAction();
            return true;
        }
    }

    return QMainWindow::eventFilter(watched, event);
}

File: app/ui/main_window.py
from __future__ import annotations

import sys

from PySide6.QtCore import QFileInfo, QProcess, QPoint, QSettings, QSize, Qt, QUrl, Signal
from PySide6.QtGui import QAction, QDesktopServices
from PySide6.QtWidgets import (
    QAbstractItemView,
    QApplication,
    QFileDialog,
    QHeaderView,
    QLabel,
    QMainWindow,
    QMenu,
    QMessageBox,
    QStatusBar,
    QTableView,
    QToolBar,
    QSplitter,
    QVBoxLayout,
    QWidget,
)

from app.models.metadata_service import MetadataService
from app.models.track import Track
from app.models.trackTableModel import TrackTableModel
from app.ui.icon_helper import app_icon
from app.ui.tag_editor_widget import TagEditorWidget


class TrackTableView(QTableView):
    filesDropped = Signal(list)

    def dragEnterEvent(self, event):
        if event.mimeData().hasUrls():
            event.acceptProposedAction()
            return
        super().dragEnterEvent(event)

    def dragMoveEvent(self, event):
        if event.mimeData().hasUrls():
            event.acceptProposedAction()
            return
        super().dragMoveEvent(event)

    def dropEvent(self, event):
        if not event.mimeData().hasUrls():
            super().dropEvent(event)
            return

        paths = [url.toLocalFile() for url in event.mimeData().urls() if url.isLocalFile()]
        if paths:
            self.filesDropped.emit(paths)
        event.acceptProposedAction()


class MainWindow(QMainWindow):
    def __init__(self):
        super().__init__()

        self._metadata_service = MetadataService()
        self._model = TrackTableModel(self)
        self._loaded_paths: set[str] = set()

        self._table = TrackTableView(self)
        self._editor = TagEditorWidget(self)
        self._recent_menu: QMenu | None = None
        self._count_label = QLabel("0 tracks")
        self._selection_label = QLabel("No track selected")

        self._open_action = QAction("Open Files...", self)
        self._open_folder_action = QAction("Open Folder...", self)
        self._clear_action = QAction("Clear Selection", self)
        self._remove_action = QAction("Remove", self)
        self._show_in_finder_action = QAction("Show in File Manager", self)
        self._quit_action = QAction("Quit", self)
        self._about_action = QAction("About TagTune", self)

        icon = app_icon()
        self.setWindowIcon(icon)
        self.setWindowTitle("TagTune")
        self.resize(1300, 900)
        self.setAcceptDrops(True)

        self._setup_ui()
        self._setup_actions()
        self._setup_toolbar()
        self._setup_menus()
        self._update_activity_bar()

    def _setup_ui(self) -> None:
        central = QWidget(self)
        layout = QVBoxLayout(central)
        layout.setContentsMargins(0, 0, 0, 0)
        splitter = QSplitter(Qt.Orientation.Vertical, central)

        self._table.setModel(self._model)
        self._table.setAcceptDrops(True)
        self._table.viewport().setAcceptDrops(True)
        self._table.setDragEnabled(False)
        self._table.setDragDropMode(QAbstractItemView.DragDropMode.DropOnly)
        self._table.setDropIndicatorShown(True)
        self._table.setDefaultDropAction(Qt.DropAction.CopyAction)
        self._table.setSelectionBehavior(QAbstractItemView.SelectionBehavior.SelectRows)
        self._table.setSelectionMode(QAbstractItemView.SelectionMode.SingleSelection)
        self._table.setAlternatingRowColors(True)
        self._table.setSortingEnabled(False)
        self._table.setContextMenuPolicy(Qt.ContextMenuPolicy.CustomContextMenu)
        self._table.customContextMenuRequested.connect(self._show_context_menu)
        self._table.filesDropped.connect(self._handle_dropped_paths)

        header = self._table.horizontalHeader()
        self._table.setIconSize(QSize(32, 32))
        self._table.verticalHeader().setVisible(False)
        self._table.verticalHeader().setDefaultSectionSize(40)
        header.setStretchLastSection(True)
        header.setSectionResizeMode(TrackTableModel.CoverColumn, QHeaderView.ResizeMode.Fixed)
        header.resizeSection(TrackTableModel.CoverColumn, 48)
        header.setSectionResizeMode(TrackTableModel.NameColumn, QHeaderView.ResizeMode.Interactive)
        header.setSectionResizeMode(TrackTableModel.ArtistColumn, QHeaderView.ResizeMode.Interactive)
        header.setSectionResizeMode(TrackTableModel.AlbumColumn, QHeaderView.ResizeMode.Interactive)
        header.setSectionResizeMode(TrackTableModel.YearColumn, QHeaderView.ResizeMode.Interactive)
        header.setSectionResizeMode(TrackTableModel.PathColumn, QHeaderView.ResizeMode.Interactive)
        header.setMinimumSectionSize(40)

        self._editor.setEditingEnabled(False)
        self._editor.saveRequested.connect(self._on_save_requested)

        splitter.addWidget(self._table)
        splitter.addWidget(self._editor)
        splitter.setStretchFactor(0, 3)
        splitter.setStretchFactor(1, 2)
        splitter.setChildrenCollapsible(False)

        layout.addWidget(splitter)
        self.setCentralWidget(central)

        self._table.selectionModel().currentRowChanged.connect(self._on_selection_changed)
        self._table.doubleClicked.connect(lambda _index: self._show_selected_in_file_manager())

        status = QStatusBar(self)
        status.addPermanentWidget(self._count_label)
        status.addPermanentWidget(self._selection_label)
        self.setStatusBar(status)
        self.statusBar().showMessage("Load audio files to begin.")

    def _setup_actions(self) -> None:
        self._open_action.triggered.connect(self._load_files)
        self._open_folder_action.triggered.connect(self._load_folder_dialog)
        self._clear_action.triggered.connect(self._clear_selection)
        self._remove_action.triggered.connect(self._remove_selected_track)
        self._show_in_finder_action.triggered.connect(self._show_selected_in_file_manager)
        self._quit_action.setMenuRole(QAction.MenuRole.QuitRole)
        self._quit_action.triggered.connect(QApplication.instance().quit)
        self._about_action.setMenuRole(QAction.MenuRole.AboutRole)
        self._about_action.triggered.connect(self._show_about)

    def _setup_toolbar(self) -> None:
        toolbar = QToolBar("Main", self)
        toolbar.setMovable(False)
        toolbar.setToolButtonStyle(Qt.ToolButtonStyle.ToolButtonTextBesideIcon)
        toolbar.addAction(self._open_action)
        toolbar.addAction(self._open_folder_action)
        toolbar.addSeparator()
        toolbar.addAction(self._clear_action)
        toolbar.addAction(self._remove_action)
        toolbar.addAction(self._show_in_finder_action)
        self.addToolBar(toolbar)

    def _setup_menus(self) -> None:
        file_menu = self.menuBar().addMenu("&File")
        file_menu.addAction(self._open_action)
        file_menu.addAction(self._open_folder_action)
        file_menu.addSeparator()
        file_menu.addAction(self._clear_action)
        file_menu.addAction(self._remove_action)
        file_menu.addAction(self._show_in_finder_action)
        file_menu.addSeparator()
        self._recent_menu = QMenu("Open Recent", self)
        file_menu.addMenu(self._recent_menu)
        file_menu.addSeparator()
        file_menu.addAction(self._quit_action)
        self._update_recent_menu()

        help_menu = self.menuBar().addMenu("&Help")
        help_menu.addAction(self._about_action)

    def _show_about(self) -> None:
        QMessageBox.about(
            self,
            "About TagTune",
            """
            <h2>TagTune</h2>
            <p><b>Version:</b> 1.0</p>
            <p>A music metadata editor built with Qt.</p>
            <p>© 2026 Retr0A</p>
            """,
        )

    def _show_message(self, message: str, timeout_ms: int = 3000) -> None:
        self.statusBar().showMessage(message, timeout_ms)

    def _update_activity_bar(self) -> None:
        self._count_label.setText(f"{self._model.rowCount()} tracks")
        current = self._table.currentIndex()
        if current.isValid():
            track = self._model.trackAt(current.row())
            if track is not None:
                self._selection_label.setText(self._selection_text(track))
                return
        self._selection_label.setText("No track selected")

    def _selection_text(self, track: Track) -> str:
        if track.title.strip():
            return track.title.strip()
        return QFileInfo(track.filePath).fileName() or "No track selected"

    def _normalize_path(self, path: str) -> str:
        info = QFileInfo(path)
        canonical = info.canonicalFilePath()
        return canonical or info.absoluteFilePath()

    def _add_to_recent(self, path: str) -> None:
        settings = QSettings()
        recent = settings.value("recentPaths", [], type=list)
        normalized = self._normalize_path(path)
        recent = [item for item in recent if item != normalized]
        recent.insert(0, normalized)
        settings.setValue("recentPaths", recent[:12])
        self._update_recent_menu()

    def _remove_from_recent(self, path: str) -> None:
        settings = QSettings()
        recent = settings.value("recentPaths", [], type=list)
        normalized = self._normalize_path(path)
        recent = [item for item in recent if item != normalized]
        settings.setValue("recentPaths", recent)
        self._update_recent_menu()

    def _update_recent_menu(self) -> None:
        if self._recent_menu is None:
            return

        self._recent_menu.clear()
        settings = QSettings()
        recent = settings.value("recentPaths", [], type=list)
        if not recent:
            self._recent_menu.setEnabled(False)
            return

        self._recent_menu.setEnabled(True)
        for path in recent:
            action = self._recent_menu.addAction(path)
            action.triggered.connect(lambda checked=False, p=path: self._open_recent_path(p))

    def _open_recent_path(self, path: str) -> None:
        info = QFileInfo(path)
        if not info.exists():
            QMessageBox.warning(self, "TagTune", f"That recent location no longer exists:\n{path}")
            self._remove_from_recent(path)
            return

        if info.isDir():
            self._load_folder(path)
            return

        if info.isFile():
            self._load_paths([path])

    def _load_folder_dialog(self) -> None:
        folder = QFileDialog.getExistingDirectory(self, "Select Music Folder")
        if not folder:
            return

        self._add_to_recent(folder)
        self._load_folder(folder)

    def _load_files(self) -> None:
        files, _ = QFileDialog.getOpenFileNames(
            self,
            "Open Audio Files",
            "",
            "Audio Files (*.mp3 *.flac *.wav *.ogg *.m4a *.aac *.mp4);;All Files (*.*)",
        )
        if not files:
            return

        self._add_to_recent(QFileInfo(files[0]).absolutePath())
        self._load_paths(files)

    def _load_folder(self, folder_path: str) -> None:
        files = self._metadata_service.scan_folder(folder_path)
        if not files:
            QMessageBox.information(self, "TagTune", f"No audio files found in {folder_path}")
            return

        added = self._load_paths(files)
        self._show_message(f"Loaded {added} file(s) from folder.", 3000)

    def _load_paths(self, paths: list[str]) -> int:
        to_load: list[str] = []
        for path in paths:
            info = QFileInfo(path)
            if not info.exists() or not info.isFile() or not self._metadata_service.is_audio_file(path):
                continue

            normalized = self._normalize_path(path)
            if normalized not in self._loaded_paths:
                to_load.append(path)

        if not to_load:
            self._show_message("No new audio files to load.", 2000)
            return 0

        loaded_tracks = self._metadata_service.load_tracks(to_load)
        added = 0
        for track in loaded_tracks:
            normalized = self._normalize_path(track.filePath)
            if normalized in self._loaded_paths:
                continue

            self._model.addTrack(track)
            self._loaded_paths.add(normalized)
            added += 1

        if added > 0 and not self._table.currentIndex().isValid():
            self._table.selectRow(0)
        elif self._table.currentIndex().isValid():
            self._sync_editor_from_selection()

        self._update_activity_bar()
        self._show_message(f"Loaded {added} file(s).", 3000)
        return added

    def _current_track(self) -> Track | None:
        index = self._table.currentIndex()
        if not index.isValid():
            return None
        return self._model.trackAt(index.row())

    def _on_selection_changed(self, _current, _previous) -> None:
        self._sync_editor_from_selection()

    def _sync_editor_from_selection(self) -> None:
        current = self._table.currentIndex()
        if not current.isValid():
            self._editor.clearTrack()
            self._editor.setEditingEnabled(False)
            self._update_activity_bar()
            return

        track = self._model.trackAt(current.row())
        if track is None:
            self._editor.clearTrack()
            self._editor.setEditingEnabled(False)
            self._update_activity_bar()
            return

        self._editor.setTrack(track)
        self._editor.setEditingEnabled(True)
        self._update_activity_bar()

    def _clear_selection(self) -> None:
        self._table.clearSelection()
        self._editor.clearTrack()
        self._editor.setEditingEnabled(False)
        self._update_activity_bar()
        self._show_message("Selection cleared.", 2000)

    def _remove_selected_track(self) -> None:
        current = self._table.currentIndex()
        if not current.isValid():
            return

        row = current.row()
        track = self._model.trackAt(row)
        if track is None:
            return

        normalized = self._normalize_path(track.filePath)
        self._loaded_paths.discard(normalized)
        removed_name = QFileInfo(track.filePath).fileName() or self._selection_text(track)

        self._model.removeTrack(row)

        row_count = self._model.rowCount()
        if row_count > 0:
            self._table.selectRow(min(row, row_count - 1))
        else:
            self._table.clearSelection()
            self._editor.clearTrack()
            self._editor.setEditingEnabled(False)

        self._sync_editor_from_selection()
        self._show_message(f"Removed {removed_name}", 2000)

    def _on_save_requested(self, track: Track) -> None:
        current = self._table.currentIndex()
        if not current.isValid():
            return

        if not self._metadata_service.save_track(track):
            QMessageBox.warning(self, "TagTune", "Unable to save track.")
            return

        self._model.updateTrack(current.row(), track)
        self._editor.setTrack(track)
        display_name = track.title.strip() or QFileInfo(track.filePath).baseName()
        self._show_message(f"Saved tags for {display_name}", 3000)

    def _show_selected_in_file_manager(self) -> None:
        track = self._current_track()
        if track is None:
            return

        file_path = QFileInfo(track.filePath)
        if not file_path.exists():
            QMessageBox.warning(self, "TagTune", f"File no longer exists:\n{track.filePath}")
            return

        absolute_path = file_path.absoluteFilePath()
        if sys.platform == "darwin":
            QProcess.startDetached("open", ["-R", absolute_path])
        elif sys.platform.startswith("win"):
            QDesktopServices.openUrl(QUrl.fromLocalFile(file_path.absolutePath()))
        else:
            QDesktopServices.openUrl(QUrl.fromLocalFile(file_path.absolutePath()))

    def _show_context_menu(self, pos: QPoint) -> None:
        index = self._table.indexAt(pos)
        menu = QMenu(self)
        if index.isValid():
            menu.addAction(self._remove_action)
            menu.addAction(self._show_in_finder_action)
        menu.exec(self._table.viewport().mapToGlobal(pos))

    def _handle_dropped_paths(self, paths: list[str]) -> None:
        file_paths: list[str] = []
        for path in paths:
            info = QFileInfo(path)
            if info.isDir():
                self._add_to_recent(path)
                self._load_folder(path)
            elif info.isFile():
                file_paths.append(path)

        if file_paths:
            self._load_paths(file_paths)
